File: nff/commands/provision.py
"""nff provision — fleet enrollment provisioning.

`nff provision batch` creates ONE shared bootstrap credential for a whole batch of devices
(DEVICE_OWNERSHIP_DESIGN.md §8). You flash the resulting credentials.h into a single firmware
image and push it to the entire batch with one OTA: every device then announces itself, shows up
in the dashboard Enroll tab, and — once accepted — automatically rolls over to a unique per-device
certificate. No per-device credential generation, no codes.
"""

import os

import click
import requests


def _fleet(fleet_url, secret):
    fleet_url = fleet_url or os.environ.get("NFF_FLEET_URL")
    secret = secret or os.environ.get("NFF_FLEET_SECRET")
    if not fleet_url:
        raise click.UsageError("fleet URL required: pass --fleet-url or set NFF_FLEET_URL")
    if not secret:
        raise click.UsageError("fleet secret required: pass --secret or set NFF_FLEET_SECRET")
    return fleet_url.rstrip("/"), secret


@click.group()
def provision():
    """Provision devices for fleet enrollment."""


@provision.command("batch")
@click.option("--project", "project_id", required=True, help="Target project id (uuid).")
@click.option("--count", type=int, default=None,
              help="Expected batch size. The server rejects enrollments beyond this as a "
                   "cloned-credential anomaly. Omit for no hard quota.")
@click.option("--out", "out_path", type=click.Path(dir_okay=False), default="credentials.h",
              help="Where to write the shared bootstrap credentials.h (default: ./credentials.h).")
@click.option("--fleet-url", default=None, help="nff-fleet base URL (or env NFF_FLEET_URL).")
@click.option("--secret", default=None, help="X-Fleet-Secret (or env NFF_FLEET_SECRET).")
def batch(project_id, count, out_path, fleet_url, secret):
    """Create a batch bootstrap credential and write its shared credentials.h.

    Build ONE firmware image with this header and flash/OTA the whole batch. The credential is
    valid for 24h; unclaimed devices are rotated automatically after that.
    """
    fleet_url, secret = _fleet(fleet_url, secret)
    body = {"project_id": project_id}
    if count is not None:
        body["count"] = count

    try:
        resp = requests.post(
            f"{fleet_url}/internal/provision-batch",
            headers={"X-Fleet-Secret": secret},
            json=body,
            timeout=30,
        )
    except requests.RequestException as exc:
        raise click.ClickException(f"could not reach fleet at {fleet_url}: {exc}")

    if resp.status_code != 200:
        detail = resp.text
        try:
            detail = resp.json().get("error", detail)
        except Exception:
            pass
        raise click.ClickException(f"fleet returned {resp.status_code}: {detail}")

    data = resp.json()
    header = data.get("bootstrap_header")
    if not header:
        raise click.ClickException("fleet did not return the bootstrap header content")

    with open(out_path, "w", encoding="utf-8") as fh:
        fh.write(header)

    click.echo(f"OK: batch {data['batch_id']} created for project {project_id}")
    click.echo(f"  credentials.h written to {out_path}")
    click.echo(f"  valid {data.get('expires_in_hours', 24)}h"
               + (f", quota {count} device(s)" if count else ", no quota"))
    click.echo("  build ONE image with this header and flash/OTA the whole batch; "
               "accept devices in the dashboard Enroll tab.")

File: nff-rs/nff/src/commands/provision.rs
//! `nff provision batch` — fleet enrollment provisioning.
//!
//! Creates ONE shared bootstrap credential for a whole batch of devices. You flash the resulting
//! credentials.h into a single firmware image and push it to the entire batch with one OTA: every
//! device announces itself, shows up in the dashboard Enroll tab, and — once accepted —
//! automatically rolls over to a unique per-device certificate. No per-device credential
//! generation, no codes.

use std::time::Duration;

use anyhow::{anyhow, Context};

use crate::cli::BatchArgs;

/// Resolve fleet URL + secret from flags, falling back to env vars.
fn resolve_fleet(
    fleet_url: Option<&str>,
    secret: Option<&str>,
) -> anyhow::Result<(String, String)> {
    let url = fleet_url
        .map(String::from)
        .or_else(|| std::env::var("NFF_FLEET_URL").ok())
        .ok_or_else(|| anyhow!("fleet URL required: pass --fleet-url or set NFF_FLEET_URL"))?;
    let secret = secret
        .map(String::from)
        .or_else(|| std::env::var("NFF_FLEET_SECRET").ok())
        .ok_or_else(|| anyhow!("fleet secret required: pass --secret or set NFF_FLEET_SECRET"))?;
    Ok((url.trim_end_matches('/').to_string(), secret))
}

pub fn run_batch(args: &BatchArgs) -> anyhow::Result<()> {
    let (fleet_url, secret) = resolve_fleet(args.fleet_url.as_deref(), args.secret.as_deref())?;

    let mut body = serde_json::json!({ "project_id": args.project });
    if let Some(count) = args.count {
        body["count"] = serde_json::json!(count);
    }

    let client = reqwest::blocking::Client::new();
    let resp = client
        .post(format!("{fleet_url}/internal/provision-batch"))
        .header("X-Fleet-Secret", secret)
        .json(&body)
        .timeout(Duration::from_secs(30))
        .send()
        .with_context(|| format!("could not reach fleet at {fleet_url}"))?;

    let status = resp.status();
    if !status.is_success() {
        let text = resp.text().unwrap_or_default();
        let detail = serde_json::from_str::<serde_json::Value>(&text)
            .ok()
            .and_then(|v| v.get("error").and_then(|e| e.as_str()).map(String::from))
            .unwrap_or(text);
        return Err(anyhow!("fleet returned {}: {detail}", status.as_u16()));
    }

    let data: serde_json::Value = resp.json().context("invalid response from fleet")?;
    let header = data
        .get("bootstrap_header")
        .and_then(|h| h.as_str())
        .ok_or_else(|| anyhow!("fleet did not return the bootstrap header content"))?;

    std::fs::write(&args.out, header)
        .with_context(|| format!("could not write {}", args.out.display()))?;

    let batch_id = data.get("batch_id").and_then(|b| b.as_str()).unwrap_or("?");
    let hours = data.get("expires_in_hours").and_then(|h| h.as_u64()).unwrap_or(24);
    println!("OK: batch {batch_id} created for project {}", args.project);
    println!("  credentials.h written to {}", args.out.display());
    match args.count {
        Some(n) => println!("  valid {hours}h, quota {n} device(s)"),
        None => println!("  valid {hours}h, no quota"),
    }
    println!(
        "  build ONE image with this header and flash/OTA the whole batch; \
         accept devices in the dashboard Enroll tab."
    );
    Ok(())
}
